### PR TITLE
Introduce team-based SP entityIDs to allow IdPs with fixed entityID to serve more than one team.

### PR DESCRIPTION
**Pull request by fisx**
Support using a single IDP with a single EntityID to set up two teams. sets up a migration, and makes teamID + EntityID unique, rather than relying on EntityID to be unique. Required to support multiple teams in environments where the IDP software cannot present anything but one EntityID (E.G.: DualShield).

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] If end-points have been added or changed: the **endpoint / config-flag checklist** (see Wire-employee only backend [wiki page](https://github.com/zinfra/backend-wiki/wiki/Checklists)) has been followed.
 - [x] If a schema migration has been added, I ran **`make git-add-cassandra-schema`** to update the cassandra schema documentation.
 - [x] **changelog.d** contains the following bits of information:
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

**Comment by fisx** [comment]
```
Failures:

  test-integration/Test/Spar/APISpec.hs:244:11: 
  1) WireIdPAPIV1.Spar.API, POST /sso/finalize-login, access granted, happy flow, responds with a very peculiar 'allowed' HTTP response
       "<?xml version=\"1.0\" encoding=\"UTF-8\"?><!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.1//EN\" \"http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd\"><html xml:lang=\"en\" xmlns=\"http://www.w3.org/1999/xhtml\"><head>  <title>wire:sso:error:forbidden</title>   <script type=\"text/javascript\">       const receiverOrigin = '*';       window.opener.postMessage({\"payload\":{\"errors\":[\"Audience mismatch: we are http://localhost:8088/sso/finalize-login, they expect one of [http://localhost:8088/sso/finalize-login/ca3474ed-3774-4185-b770-65a20ecb5220]\"],\"label\":\"forbidden\"},\"type\":\"AUTH_ERROR\"}, receiverOrigin);   </script></head></html>" does not contain "<title>wire:sso:success</title>"
```

The most interesting part is this:

```
["Audience mismatch: we are http://localhost:8088/sso/finalize-login, they expect one of [http://localhost:8088/sso/finalize-login/ca3474ed-3774-4185-b770-65a20ecb5220]"]
```

So for IdPs expecting the old behavior, the auth request names the new finalize-login path.  That shouldn't be too hard to fix, no?

File: libs/wire-api/src/Wire/API/User/IdentityProvider.hs
Comment (fisx) [reply libs/wire-api/src/Wire/API/User/IdentityProvider.hs:69]: not in this PR, no time!

File: services/spar/src/Spar/API.hs
Comment (fisx) [reply services/spar/src/Spar/API.hs]: 178528fd9aec63c80150bc5006cac9db4db3f75f

File: services/spar/src/Spar/API.hs
Comment (fisx) [reply services/spar/src/Spar/API.hs:359]: 5ede16ff2290febbacc77e19a563301ca45e0d4c

File: services/spar/src/Spar/API.hs
Comment (fisx) [reply services/spar/src/Spar/API.hs:363]: in fact, this is not redundant, but wrong.  instead of `idp'`, we should look at the version of the idp about to be replaced.  that could conceivable be v1, while we're replacing it with a v2, and that's not supported because i haven't thought it through.

**Review by julialongtin — APPROVED** [review]
reviewed over wire. was fun!

**Comment by jschaul** [comment]
There are 4 failing tests (consistently across two CI runs):

```
Failures:


  test-integration/Test/Spar/APISpec.hs:138:7: 

  1) WireIdPAPIV1.Spar.API.metadata metadata (legacy)

       predicate failed on: Response {responseStatus = Status {statusCode = 200, statusMessage = "OK"}, responseVersion = HTTP/1.1, responseHeaders = [("Transfer-Encoding","chunked"),("Date","Mon, 13 Sep 2021 19:23:48 GMT"),("Server","Warp/3.3.13"),("Cache-Control","no-cache, no-store"),("Pragma","no-cache"),("Content-Type","application/xml")], responseBody = Just "<md:EntityDescriptor xmlns:samlp=\"urn:oasis:names:tc:SAML:2.0:protocol\" xmlns:samla=\"urn:oasis:names:tc:SAML:2.0:assertion\" xmlns:samlm=\"urn:oasis:names:tc:SAML:2.0:metadata\" xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\" entityID=\"http://spar:8080/sso/finalize-login\" xmlns:md=\"urn:oasis:names:tc:SAML:2.0:metadata\"><md:SPSSODescriptor ID=\"_bf1b0086-392f-4a30-8380-71adca8e1471\" WantAssertionsSigned=\"true\" cacheDuration=\"PT2592000S\" protocolSupportEnumeration=\"urn:oasis:names:tc:SAML:2.0:protocol\" validUntil=\"2022-09-13T19:23:48.7276188Z\"><md:Organization><md:OrganizationName xml:lang=\"EN\">spar</md:OrganizationName><md:OrganizationDisplayName xml:lang=\"EN\">spar</md:OrganizationDisplayName><md:OrganizationURL xml:lang=\"EN\">http://spar:8080/sso/finalize-login</md:OrganizationURL></md:Organization><md:ContactPerson contactType=\"support\"><md:Company>Example Company</md:Company><md:EmailAddress>email:backend+spar@wire.com</md:EmailAddress></md:ContactPerson><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:entity</md:NameIDFormat><md:AssertionConsumerService Binding=\"urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST\" Location=\"http://spar:8080/sso/finalize-login\" index=\"0\" isDefault=\"true\"/></md:SPSSODescriptor></md:EntityDescriptor>", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}


  To rerun use: --match "/WireIdPAPIV1/Spar.API/metadata/metadata (legacy)/"


  test-integration/Test/Spar/APISpec.hs:150:7: 

  2) WireIdPAPIV1.Spar.API.metadata metadata (with team-id)

       predicate failed on: Response {responseStatus = Status {statusCode = 200, statusMessage = "OK"}, responseVersion = HTTP/1.1, responseHeaders = [("Transfer-Encoding","chunked"),("Date","Mon, 13 Sep 2021 19:23:48 GMT"),("Server","Warp/3.3.13"),("Cache-Control","no-cache, no-store"),("Pragma","no-cache"),("Content-Type","application/xml")], responseBody = Just "<md:EntityDescriptor xmlns:samlp=\"urn:oasis:names:tc:SAML:2.0:protocol\" xmlns:samla=\"urn:oasis:names:tc:SAML:2.0:assertion\" xmlns:samlm=\"urn:oasis:names:tc:SAML:2.0:metadata\" xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\" entityID=\"http://spar:8080/sso/finalize-login/3f1a01e2-1092-11ec-846d-3344b58840fe\" xmlns:md=\"urn:oasis:names:tc:SAML:2.0:metadata\"><md:SPSSODescriptor ID=\"_0665b2c1-ea4b-441d-8987-fc9ebcec0dc1\" WantAssertionsSigned=\"true\" cacheDuration=\"PT2592000S\" protocolSupportEnumeration=\"urn:oasis:names:tc:SAML:2.0:protocol\" validUntil=\"2022-09-13T19:23:48.7315704Z\"><md:Organization><md:OrganizationName xml:lang=\"EN\">spar</md:OrganizationName><md:OrganizationDisplayName xml:lang=\"EN\">spar</md:OrganizationDisplayName><md:OrganizationURL xml:lang=\"EN\">http://spar:8080/sso/finalize-login/3f1a01e2-1092-11ec-846d-3344b58840fe</md:OrganizationURL></md:Organization><md:ContactPerson contactType=\"support\"><md:Company>Example Company</md:Company><md:EmailAddress>email:backend+spar@wire.com</md:EmailAddress></md:ContactPerson><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:entity</md:NameIDFormat><md:AssertionConsumerService Binding=\"urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST\" Location=\"http://spar:8080/sso/finalize-login/3f1a01e2-1092-11ec-846d-3344b58840fe\" index=\"0\" isDefault=\"true\"/></md:SPSSODescriptor></md:EntityDescriptor>", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}


  To rerun use: --match "/WireIdPAPIV1/Spar.API/metadata/metadata (with team-id)/"


  test-integration/Test/Spar/APISpec.hs:138:7: 

  3) WireIdPAPIV2.Spar.API.metadata metadata (legacy)

       predicate failed on: Response {responseStatus = Status {statusCode = 200, statusMessage = "OK"}, responseVersion = HTTP/1.1, responseHeaders = [("Transfer-Encoding","chunked"),("Date","Mon, 13 Sep 2021 19:25:52 GMT"),("Server","Warp/3.3.13"),("Cache-Control","no-cache, no-store"),("Pragma","no-cache"),("Content-Type","application/xml")], responseBody = Just "<md:EntityDescriptor xmlns:samlp=\"urn:oasis:names:tc:SAML:2.0:protocol\" xmlns:samla=\"urn:oasis:names:tc:SAML:2.0:assertion\" xmlns:samlm=\"urn:oasis:names:tc:SAML:2.0:metadata\" xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\" entityID=\"http://spar:8080/sso/finalize-login\" xmlns:md=\"urn:oasis:names:tc:SAML:2.0:metadata\"><md:SPSSODescriptor ID=\"_983969b0-7bdb-428a-9ef4-99ec5b376a24\" WantAssertionsSigned=\"true\" cacheDuration=\"PT2592000S\" protocolSupportEnumeration=\"urn:oasis:names:tc:SAML:2.0:protocol\" validUntil=\"2022-09-13T19:25:53.0600829Z\"><md:Organization><md:OrganizationName xml:lang=\"EN\">spar</md:OrganizationName><md:OrganizationDisplayName xml:lang=\"EN\">spar</md:OrganizationDisplayName><md:OrganizationURL xml:lang=\"EN\">http://spar:8080/sso/finalize-login</md:OrganizationURL></md:Organization><md:ContactPerson contactType=\"support\"><md:Company>Example Company</md:Company><md:EmailAddress>email:backend+spar@wire.com</md:EmailAddress></md:ContactPerson><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:entity</md:NameIDFormat><md:AssertionConsumerService Binding=\"urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST\" Location=\"http://spar:8080/sso/finalize-login\" index=\"0\" isDefault=\"true\"/></md:SPSSODescriptor></md:EntityDescriptor>", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}


  To rerun use: --match "/WireIdPAPIV2/Spar.API/metadata/metadata (legacy)/"


  test-integration/Test/Spar/APISpec.hs:150:7: 

  4) WireIdPAPIV2.Spar.API.metadata metadata (with team-id)

       predicate failed on: Response {responseStatus = Status {statusCode = 200, statusMessage = "OK"}, responseVersion = HTTP/1.1, responseHeaders = [("Transfer-Encoding","chunked"),("Date","Mon, 13 Sep 2021 19:25:52 GMT"),("Server","Warp/3.3.13"),("Cache-Control","no-cache, no-store"),("Pragma","no-cache"),("Content-Type","application/xml")], responseBody = Just "<md:EntityDescriptor xmlns:samlp=\"urn:oasis:names:tc:SAML:2.0:protocol\" xmlns:samla=\"urn:oasis:names:tc:SAML:2.0:assertion\" xmlns:samlm=\"urn:oasis:names:tc:SAML:2.0:metadata\" xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\" entityID=\"http://spar:8080/sso/finalize-login/3f1a01e2-1092-11ec-846d-3344b58840fe\" xmlns:md=\"urn:oasis:names:tc:SAML:2.0:metadata\"><md:SPSSODescriptor ID=\"_b488dd1e-a95e-4b73-8e28-14a5d8c4c173\" WantAssertionsSigned=\"true\" cacheDuration=\"PT2592000S\" protocolSupportEnumeration=\"urn:oasis:names:tc:SAML:2.0:protocol\" validUntil=\"2022-09-13T19:25:53.0641463Z\"><md:Organization><md:OrganizationName xml:lang=\"EN\">spar</md:OrganizationName><md:OrganizationDisplayName xml:lang=\"EN\">spar</md:OrganizationDisplayName><md:OrganizationURL xml:lang=\"EN\">http://spar:8080/sso/finalize-login/3f1a01e2-1092-11ec-846d-3344b58840fe</md:OrganizationURL></md:Organization><md:ContactPerson contactType=\"support\"><md:Company>Example Company</md:Company><md:EmailAddress>email:backend+spar@wire.com</md:EmailAddress></md:ContactPerson><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:entity</md:NameIDFormat><md:AssertionConsumerService Binding=\"urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST\" Location=\"http://spar:8080/sso/finalize-login/3f1a01e2-1092-11ec-846d-3344b58840fe\" index=\"0\" isDefault=\"true\"/></md:SPSSODescriptor></md:EntityDescriptor>", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}


  To rerun use: --match "/WireIdPAPIV2/Spar.API/metadata/metadata (with team-id)/"


Randomized with seed 2095268154


Finished in 253.0226 seconds

507 examples, 4 failures, 72 pending
```

**Comment by fisx** [comment]
(at least some of) the failures are happening here:

```
- "...Location=\"http://spar:8080/sso/finalize-login\"..."
+ "...Location=\"http://localhost:8088/sso/finalize-login\"..."
```

That's why I couldn't reproduce this locally.  I'll think of a cheap way to abstract over the spar hostname.

File: services/spar/test-integration/Util/Types.hs
Comment (fisx) [reply services/spar/test-integration/Util/Types.hs:124]: next PR.

File: services/spar/src/Spar/Data.hs
Comment (fisx) [reply services/spar/src/Spar/Data.hs:610]: next PR.

Comment (fisx) [reply services/spar/src/Spar/Data.hs:610]: (shouldn't make a difference; no idp is ever stored in both tables.)

Comment (fisx) [reply services/spar/src/Spar/Data.hs:610]: > (shouldn't make a difference except in performance; no idp is ever stored in both tables.)

File: services/spar/src/Spar/App.hs
Comment (fisx) [reply services/spar/src/Spar/App.hs:240]: not in this PR!

File: services/spar/test-integration/Util/Types.hs
Comment (fisx) [reply services/spar/test-integration/Util/Types.hs:124]: next PR!

File: services/spar/src/Spar/Data.hs
Comment (fisx) [reply services/spar/src/Spar/Data.hs]: b9498f42d084c38609c0a9169a449d0c06c0dee3

File: services/spar/src/Spar/Data.hs
Comment (fisx) [reply services/spar/src/Spar/Data.hs]: b9498f42d084c38609c0a9169a449d0c06c0dee3

File: services/spar/src/Spar/Data.hs
Comment (fisx) [reply services/spar/src/Spar/Data.hs]: b9498f42d084c38609c0a9169a449d0c06c0dee3

File: services/spar/src/Spar/Data.hs
Comment (fisx) [reply services/spar/src/Spar/Data.hs]: 8cad1b2cd28bbbefcaabe6fda3b27159b0bf712b

File: services/spar/src/Spar/App.hs
Comment (fisx) [reply services/spar/src/Spar/App.hs]: 79488d5a824d6bdbc7fdc949db82c2c5294b7cbe

File: services/spar/src/Spar/API.hs
Comment (fisx) [reply services/spar/src/Spar/API.hs]: 379c56c865ff36e3f5112882d5217bf26be17565

**Comment by fisx** [comment]
I also think this is ready to merge now, if concourse is done.  But @julialongtin I wouldn't mind a second opinion on 379c56c865ff36e3f5112882d5217bf26be17565...

File: services/spar/src/Spar/API.hs
Comment (fisx) [inline services/spar/src/Spar/API.hs:372]: i think this line is redundant, but that's not affecting behavior, so let's fix that in a different PR.

Comment (fisx) [reply services/spar/src/Spar/API.hs:372]: we should also clarify: if this is a replacement, then the issuer of the replacee is different, so we wouldn't find the replacee here.

if we want to replace an idp with one that has the same issuer, we would use PUT instead of POST.

File: services/spar/src/Spar/App.hs
Comment (fisx) [reply services/spar/src/Spar/App.hs:240]: d51f1404031c8713f8a69202828192019ebeba5b

File: services/spar/src/Spar/Data.hs
Comment (fisx) [reply services/spar/src/Spar/Data.hs:603]: 4f295ace138b5080ddbf5807d2546abf88436505

File: services/spar/src/Spar/API.hs
Comment (fisx) [reply services/spar/src/Spar/API.hs:372]: 28dba9c410809a9e20d660de84e488f67b799145, e2fdfdf4d1f9f6bab338dc360360a6d07fed8f5f

File: libs/wire-api/src/Wire/API/User/IdentityProvider.hs
Comment (fisx) [reply libs/wire-api/src/Wire/API/User/IdentityProvider.hs:69]: c49207f701dcb0861167999011e24fa635726a2c

File: services/spar/test-integration/Util/Types.hs
Comment (fisx) [reply services/spar/test-integration/Util/Types.hs:124]: 5a3d713e719f0cf78b943f7af3ba3a33518f0fb8

`make -C services/spar i` takes currently takes 3m22.476s on my laptop, including all the uninteresting tests.  So this is the maximum time improvement that we can get out of this.

**Pull request merged**
